src/prefixtrie.cpp

```diff
@@ -0,0 +1,486 @@
+void CPrefixTrie<TKey, TData>::erase(const TKey& key, std::shared_ptr<Node>& node)
+{
+    std::vector<typename TChildren::value_type> nodes;
+    nodes.emplace_back(TKey(), node);
+    using CBType = callback<std::shared_ptr<Node>>;
+    CBType cb = [&nodes](const TKey& k, std::shared_ptr<Node> n) {
+        nodes.emplace_back(k, n);
+    };
+    if (!find(key, node, cb))
+        return;
+
+    nodes.back().second->data = std::make_shared<TData>();
+    for (; nodes.size() > 1; nodes.pop_back()) {
+        // if we have only one child and no data ourselves, bring them up to our level
+        auto& cNode = nodes.back().second;
+        auto onlyOneChild = cNode->children.size() == 1;
+        auto noData = cNode->data->empty();
+        if (onlyOneChild && noData) {
+            auto child = cNode->children.begin();
+            auto& prefix = nodes.back().first;
+            auto newKey = prefix;
+            auto& postfix = child->first;
+            newKey.insert(newKey.end(), postfix.begin(), postfix.end());
+            auto& pNode = nodes[nodes.size() - 2].second;
+            pNode->children.emplace(newKey, std::move(child->second));
+            pNode->children.erase(prefix);
+            --size;
+            continue;
+        }
+
+        auto noChildren = cNode->children.empty();
+        if (noChildren && noData) {
+            auto& pNode = nodes[nodes.size() - 2].second;
+            pNode->children.erase(nodes.back().first);
+            --size;
+            continue;
+        }
+        break;
+    }
+}
+
+template <typename TKey, typename TData>
+CPrefixTrie<TKey, TData>::CPrefixTrie() : size(0), root(std::make_shared<Node>())
+{
+    root->data = std::make_shared<TData>();
+}
+
+template <typename TKey, typename TData>
+template <typename TDataUni>
+typename CPrefixTrie<TKey, TData>::iterator CPrefixTrie<TKey, TData>::insert(const TKey& key, TDataUni&& data)
+{
+    auto& node = key.empty() ? root : insert(key, root);
+    node->data = std::make_shared<TData>(std::forward<TDataUni>(data));
+    return key.empty() ? begin() : iterator{key, node};
+}
+
+template <typename TKey, typename TData>
+typename CPrefixTrie<TKey, TData>::iterator CPrefixTrie<TKey, TData>::copy(CPrefixTrie<TKey, TData>::const_iterator it)
+{
+    auto& key = it.key();
+    auto& node = key.empty() ? root : insert(key, root);
+    node->data = it.node.lock()->data;
+    return key.empty() ? begin() : iterator{key, node};
+}
+
+template <typename TKey, typename TData>
+template <typename TDataUni>
+typename CPrefixTrie<TKey, TData>::iterator CPrefixTrie<TKey, TData>::insert(CPrefixTrie<TKey, TData>::iterator& it, const TKey& key, TDataUni&& data)
+{
+    auto shared = it.node.lock();
+    assert(shared);
+    auto copy = it;
+    if (!key.empty()) {
+        auto name = it.key();
+        name.insert(name.end(), key.begin(), key.end());
+        auto& node = insert(key, shared);
+        copy = iterator{name, node};
```

Comment:
Not a true forward iterator -- that would require the full stack. This is a children-only iterator. It's true of all the places we create the iterator here. We are specifically choosing to not initialize them fully, and we're using the weak_ptr workaround for the place where we need the parent function but don't have it.

Reply:
That's by design, iterator is forward through children, i mean this is not an exception. Thus we don't have ability to obtain parent, that's nodes stand for. We can go for *full* iterator but i'm afraid we will fill iterator stack for nothing i.e. we will not use it.